### PR TITLE
Support AWS KMS credentials using decryption secretRef

### PR DESCRIPTION
**Pull request by aryan9600**
Adds support for using `.spec.decryption` to refer to custom AWS KMS credentials and override global decryption configuration.

Signed-off-by: Sanskar Jaiswal <sanskar.jaiswal@weave.works>

### CODE REVIEW

**Review by stefanprodan — COMMENTED** [review]
Please extend the documentation to cover AWS KMS:

- Explain static keys here https://github.com/fluxcd/kustomize-controller/blob/main/docs/spec/v1beta2/kustomization.md#decryption-secret-reference
- Explain IRSA here https://github.com/fluxcd/kustomize-controller/blob/main/docs/spec/v1beta2/kustomization.md#controller-global-decryption

PS. For IRSA you can take some of the info from https://fluxcd.io/docs/guides/mozilla-sops/#encrypting-secrets-using-various-cloud-providers

**Comment by stefanprodan** [comment]
@aryan9600 it's fine to link to the guide, but we need a section like this one https://github.com/fluxcd/kustomize-controller/blob/main/docs/spec/v1beta2/kustomization.md#gcp-kms for AWS.

**Comment by hiddeco** [comment]
There is no documentation which instructs how to supply a SecretRef and what data the Secret must contain for AWS. Which is explained in detail for any other custom implementation we have.

**Comment by aryan9600** [comment]
Yep, I'm working on that, I was talking about the ISRA stuff 😅

File: internal/sops/keyservice/server.go
Comment (hiddeco) [inline internal/sops/keyservice/server.go:106]: Same [suggestion as for GCP KMS](https://github.com/fluxcd/kustomize-controller/pull/635#discussion_r856353043).

File: internal/sops/awskms/keysource_test.go
Comment (hiddeco) [inline internal/sops/awskms/keysource_test.go:26]: Instead of copying the tests from upstream, please rewrite them to get the same coverage as the other keyservice implementations, while also [including `_Compat` tests](https://github.com/fluxcd/kustomize-controller/pull/635#discussion_r861344121).

File: internal/sops/awskms/keysource.go
Comment (hiddeco) [inline internal/sops/awskms/keysource.go:27]: There is a newer SDK available: https://github.com/aws/aws-sdk-go-v2

Comment (aryan9600) [reply internal/sops/awskms/keysource.go:27]: yes, but upstream sops [still use v1](https://github.com/mozilla/sops/blob/master/kms/keysource.go#L17), so i thought it might be better to use that as well to avoid any incompatibility risks 😓

Comment (hiddeco) [reply internal/sops/awskms/keysource.go:27]: That's what the `_Compat` tests are for, the same is true for the Azure KMS integration :-)

**Comment by aryan9600** [comment]
Fuzzing tests are failing because of invalid fuzzing tests in `aws-sdk-go-v2`, ref: https://github.com/aws/aws-sdk-go-v2/issues/1693

File: internal/sops/awskms/keysource.go
Comment (hiddeco) [inline internal/sops/awskms/keysource.go:39]: Think it would help here if at least one match example would be given per rule.

File: internal/sops/awskms/keysource.go
Comment (hiddeco) [inline internal/sops/awskms/keysource.go:45]: How is this implementation different from upstream?

File: internal/sops/awskms/keysource.go
Comment (hiddeco) [inline internal/sops/awskms/keysource.go:67]:
```suggestion
// NewCredsProvider returns a Creds object with the provided aws.CredentialsProvider.
```

File: internal/sops/awskms/keysource.go
Comment (hiddeco) [inline internal/sops/awskms/keysource.go:96]:
```suggestion
// EncryptedDataKey returns the encrypted data key this master key holds.
```

File: internal/sops/awskms/keysource.go
Comment (hiddeco) [inline internal/sops/awskms/keysource.go:101]:
```suggestion
// SetEncryptedDataKey sets the encrypted data key for this master key.
```

File: internal/sops/awskms/keysource.go
Comment (hiddeco) [inline internal/sops/awskms/keysource.go:106]:
```suggestion
// Encrypt takes a SOPS data key, encrypts it with KMS and stores the result in
// the EncryptedKey field.
```

File: internal/sops/awskms/keysource.go
Comment (hiddeco) [inline internal/sops/awskms/keysource.go:161]:
```suggestion
// ToString converts the key to a string representation.
```

File: internal/sops/awskms/keysource.go
Comment (hiddeco) [inline internal/sops/awskms/keysource.go:166]:
```suggestion
// NewMasterKey creates a new MasterKey from an ARN, role and context, setting the
// creation date to the current date.
```

File: internal/sops/awskms/keysource.go
Comment (hiddeco) [inline internal/sops/awskms/keysource.go:176]:
```suggestion
// NewMasterKeyFromArn takes an ARN string and returns a new MasterKey for that
// ARN.
```

File: internal/sops/awskms/keysource.go
Comment (hiddeco) [inline internal/sops/awskms/keysource.go:193]: Feel this could use a comment that if `key.credentialsProvider` is `nil`, the behavior of `LoadDefaultConfig` is to look at the environment.

File: internal/sops/awskms/keysource.go
Comment (hiddeco) [inline internal/sops/awskms/keysource.go:81]:
```suggestion
// LoadCredsProviderFromYAML parses the given YAML returns a CredsProvider object
// which contains the credentials provider used for authenticating towards AWS KMS.
func LoadCredsProviderFromYaml(b []byte) (*CredsProvider, error) {
```

(the package name is already context defining).

File: internal/sops/awskms/keysource.go
Comment (hiddeco) [inline internal/sops/awskms/keysource.go:125]:
```suggestion
// EncryptIfNeeded encrypts the provided SOPS data key, if it has not been
// encrypted yet.
```

File: internal/sops/awskms/keysource.go
Comment (hiddeco) [inline internal/sops/awskms/keysource.go:212]: What does this do?

**Review by hiddeco — COMMENTED** [review]
This is shaping up well, couple of nitpicks mostly around making our lives easier when we look at this again in a couple months :smile:

File: controllers/kustomization_decryptor.go
Comment (hiddeco) [inline controllers/kustomization_decryptor.go:134]:
```suggestion
	// awsCredsProvider is the AWS credentials provider object used to authenticate
	// towards any AWS KMS.
	awsCredsProvider *awskms.CredsProvider
```

File: docs/spec/v1beta2/kustomization.md
Comment (stefanprodan) [inline docs/spec/v1beta2/kustomization.md:1273]: Please remove the labels, this is needed for Azure not AWS.

Comment (aryan9600) [reply docs/spec/v1beta2/kustomization.md:1273]: copy pasted the yaml, the labels sneaked past me

File: internal/sops/awskms/keysource.go
Comment (hiddeco) [inline internal/sops/awskms/keysource.go:39]:
```suggestion
	// arnRegex matches an AWS ARN.
	// Example: arn:aws:kms:us-west-2:107501996527:key/612d5f0p-p1l3-45e6-aca6-a5b005693a48
	arnRegex = `^arn:aws[\w-]*:kms:(.+):[0-9]+:(key|alias)/.+$`
```

File: internal/sops/awskms/keysource.go
Comment (hiddeco) [inline internal/sops/awskms/keysource.go:41]:
```suggestion
	// stsSessionRegex matches an AWS STS session name.
	// Examples: john_s, sops@42WQm042
	stsSessionRegex = "[^a-zA-Z0-9=,.@-_]+"
```

File: internal/sops/awskms/keysource.go
Comment (hiddeco) [inline internal/sops/awskms/keysource.go:67]:
```suggestion
// CredsProvider is a wrapper around aws.CredentialsProvider used for authenticating
// towards AWS KMS.
```

File: internal/sops/awskms/keysource.go
Comment (hiddeco) [inline internal/sops/awskms/keysource.go:204]:
```suggestion
	// from the environment.
```

File: internal/sops/awskms/keysource.go
Comment (hiddeco) [inline internal/sops/awskms/keysource.go:255]: This should be above private methods.

File: internal/sops/awskms/keysource.go
Comment (hiddeco) [inline internal/sops/awskms/keysource.go:63]:
```suggestion
	// epResolver can be used to override the endpoint that the AWS client
	// resolves to by default. This is mostly used for testing purposes as it can
	// not be injected using e.g. an environment variable. The field is not
	// publicly exposed, nor configurable.
	epResolver aws.EndpointResolver
```

File: internal/sops/keyservice/options.go
Comment (hiddeco) [inline internal/sops/keyservice/options.go:60]:
```suggestion
// WithAWSKeys configures the AWS credentials on the Server.
```

File: internal/sops/keyservice/server.go
Comment (hiddeco) [inline internal/sops/keyservice/server.go:49]: This is no longer correct.

File: internal/sops/awskms/keysource.go
Comment (hiddeco) [inline internal/sops/awskms/keysource.go:57]: Missing docs, especially `EncryptionContext` is not precisely clear.

**Review by hiddeco — APPROVED** [review]
Stellar job :100:

File: docs/spec/v1beta2/kustomization.md
Comment (pjbgf) [inline docs/spec/v1beta2/kustomization.md:1258]:
```suggestion
Furthermore, you can also use the usual [environment variables used for specifying AWS
```

File: internal/sops/awskms/keysource.go
Comment (pjbgf) [inline internal/sops/awskms/keysource.go:175]:
```suggestion
	return time.Since(key.CreationDate) > (kmsTTL)
```

**Review by pjbgf — COMMENTED** [review]
two tiny nits, apart from that LGTM

great work @aryan9600!

**Comment by stefanprodan** [comment]
What are going to do about fuzzing? We can't merge this and have all PRs from now on failing. Can we disable the AWS tests?

**Comment by pjbgf** [comment]
> What are going to do about fuzzing? We can't merge this and have all PRs from now on failing. Can we disable the AWS tests?

We will disable them for the time being whilst we wait for the fix upstream. The approach we discussed so far is picking a different conditional tag for our own fuzz tests.

File: docs/spec/v1beta2/kustomization.md
Comment (pjbgf) [inline docs/spec/v1beta2/kustomization.md:1302]:
```suggestion
> **Note:**: If you're mounting a secret containing the AWS credentials as a file in the `kustomize-controller` pod,
```
```suggestion
> **Note:**: If you're mounting a secret containing the AWS credntials as a file in the `kustomize-controller` pod,
```

Comment (pjbgf) [inline docs/spec/v1beta2/kustomization.md:1302]:
```suggestion
> **Note:**: If you're mounting a secret containing the AWS credentials as a file in the `kustomize-controller` pod,
```
```suggestion
> **Note:**: If you're mounting a secret containing the AWS credntials as a file in the `kustomize-controller` pod,
```

**Review by pjbgf — APPROVED** [review]
Great job @aryan9600! :clap: :clap:

LGTM

**Comment by hiddeco** [comment]
Can we please change the license on this to MPL before we merge.

**Review by hiddeco — APPROVED** [review]

**Pull request merged**
